alhimick-prog/hometask-3/rainfall.rb

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+INPUT_PROMPT = 'Enter city name:'
+ERROR_BLANK_STRING = 'City name can not be blank!'
+ERROR_FILE_READ_FAIL = 'Error: data file not found'
+ERROR_MOUNT_NUMBER = 'Error: something went wrong when counting months.'
+MEAN = 'Rainfall mean:'
+VARIANCE = 'Rainfall variance:'
+EXIT_STRING = 'exit!'
+TOWN_ERROR = -1
+MOUNTH_NUMBER = 12
+CURRENT_FILE = 'data.txt'
+
+def file_reader(filename)
+  if File.exist?(filename)
+    file = File.new(filename, 'r')
+    file.sysread(file.size)
+    # I know i could use a method .readlines to get an array from file,
+    # but in the condition of assignment
+    # data is passed to methods 'mean' and 'variance' as a string.
+  else
+    puts ERROR_FILE_READ_FAIL
+    exit
+  end
+end
+
+def town_pos(town, strng)
+  pos = /(^|\s)#{town}:/ =~ strng
+  case pos
+  when /^#{town}:/ =~ strng
+    pos
+  when /\s#{town}:/ =~ strng
+    pos + 1
+  else
+    false
+  end
+end
+
+def form_substring(pos, strng)
+  substr = ''
+  loop do
+    substr += strng[pos]
+    pos += 1
+    break if strng[pos] == "\n" || pos == strng.size
+  end
+  substr
+end
+
+def form_array(town, strng)
+  if town_pos(town, strng)
+    start_search_pos = town_pos(town, strng) + town.size + 1
+    substr = form_substring(start_search_pos, strng)
+    substr.scan(/\d+\.*\d*/).map(&:to_f)
+  else
+    false
+  end
+end
+
+def mean(town, strng)
+  array = form_array(town, strng)
+  return TOWN_ERROR unless array
+
+  puts ERROR_MOUNT_NUMBER if array.size != MOUNTH_NUMBER
+  array.sum / array.size
```

Comment:
@alhimick-prog you print error but continue to calculate `mean`. I think we need to stop the program with error, right?

Reply:
@lincAlinc in this task, this check of quantity is not required at all, since everything works correctly and with verified data. I added this check in case of working with other data sources that may be incorrectly processed or may themselves be incorrect (for example, the line contains data not for all months). In this case, the user will receive a warning, but the program will continue to run. Perhaps this check is not needed?

Reply:
@alhimick-prog if you want just add warning, you need to rename this `ERROR_MOUNT_NUMBER`